### PR TITLE
Update metrics tracker to most recent specifications

### PR DESCRIPTION
**Pull request by defagos**
# Description

This PR updates our metrics tracker implementation so that it conforms to the latest version of our specifications. After having identified a few issues during the implementation I also realized that having no UTs for our metrics tracker implementation was risky. I therefore introduced the necessary testing infrastructure so that we can ensure our implementation stays reliable over time.

# Changes made

- Avoid `stop` event being sent when no `start` event has been sent.
- Use metric event date as date sent to the monitoring service.
- Use lower-case UUIDs which are more [standard](https://www.itu.int/rec/T-REC-X.667), improving consistency with UUIDs delivered by Android and the web.
- Send bitrate and bandwidth in bits per second.
- Add schema version.
- Group stall information.
- Rename key for storage of QoE metrics.
- Implement heartbeats.
- Revisit metrics tracker configuration to support optional heartbeat interval customization.
- Introduce testing infrastructure and improve consistency with the similar formalism existing for analytics tracker testing.
- Fix `PillarboxMonitoring` scheme.
- Fix MB calculation.
- Cache device type in comScore analytics.
- Group files in a better way in Monitoring and Analytics packages.

# Checklist

- [x] APIs have been properly documented (if relevant).
- [x] The documentation has been updated (if relevant).
- [x] New unit tests have been written (if relevant).
- [x] The demo has been updated (if relevant).

### CODE REVIEW

File: Tests/MonitoringTests/MetricHitExpectation.swift [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+@testable import PillarboxMonitoring
+
+private struct _MetricHitExpectation<Data>: MetricHitExpectation where Data: Encodable {
+    let eventName: EventName
+    private let evaluate: (MetricPayload<Data>) -> Void
+
+    fileprivate init(eventName: EventName, evaluate: @escaping (MetricPayload<Data>) -> Void) {
+        self.eventName = eventName
+        self.evaluate = evaluate
+    }
```
Comment (waliid) [inline Tests/MonitoringTests/MetricHitExpectation.swift:16]:
```suggestion
    init(eventName: EventName, evaluate: @escaping (MetricPayload<Data>) -> Void) {
        self.eventName = eventName
        self.evaluate = evaluate
    }
```

Unlike `ComScoreHitExpectation` and `CommandersActHitExpectation` this object is private, so we don't need a `fileprivate` initializer.

**Review by waliid — APPROVED** [review]

File: Sources/Monitoring/Types/MetricStartData.swift [hunk 2]
```diff
@@ -40,13 +30,18 @@ extension MetricStartData {
...
+    }
+
+    struct Player: Encodable {
+        let name: String
+        let platform: String
+        let version: String
     }
 
-    struct TimeMetrics: Encodable {
+    struct QoEMetrics: Encodable {
```
[the remark this answers is not in the document]
Comment (defagos) [reply Sources/Monitoring/Types/MetricStartData.swift:44]: Sure, will do

File: Sources/Monitoring/Types/Notification.Name.swift [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+//  Copyright (c) SRG SSR. All rights reserved.
+//
+//  License information is available from the LICENSE file.
+//
+
+import Foundation
+
+enum MetricRequestInfoKey: String {
+    case identifier = "MetricRequestIdentifier"
+    case payload = "MetricRequestIdentifierPayload"
```
[the remark this answers is not in the document]
Comment (defagos) [reply Sources/Monitoring/Types/Notification.Name.swift:11]: These are keys in a user info dictionary. If I am not mistaken we have a similar convention for seek notifications, but I'll have another look.

File: Sources/Monitoring/Types/MetricStartData.swift (hunk 2 above)
[the remark this answers is not in the document]
Comment (defagos) [reply Sources/Monitoring/Types/MetricStartData.swift]: Done.

File: Sources/Monitoring/Types/Notification.Name.swift (hunk 3 above)
[the remark this answers is not in the document]
Comment (defagos) [reply Sources/Monitoring/Types/Notification.Name.swift:11]: If you take a look at the strings associated with notifications / keys in `AVFoundation` at random you can see that Camel case is used:

```
(lldb) po AVPlayerInterstitialEventMonitorAssetListResponseStatusDidChangeNotification
AssetListResponseStatusDidChangeNotification

(lldb) po AVPlayerItemFailedToPlayToEndTimeErrorKey
AVPlayerItemFailedToPlayToEndTimeErrorKey
```

So I guess sticking with Camel case is quite consistent here.

File: Tests/MonitoringTests/MetricHitExpectation.swift (hunk 1 above)
Comment (defagos) [reply Tests/MonitoringTests/MetricHitExpectation.swift:16]: Remove superfluous scope.

File: Sources/Monitoring/Types/Notification.Name.swift (hunk 3 above)
[the remark this answers is not in the document]
Comment (waliid) [reply Sources/Monitoring/Types/Notification.Name.swift:11]: Alright, thanks for the investigation.

**Review by waliid — APPROVED** [review]

**Comment by defagos** [comment]
Thank you very much for the review!

**Pull request merged**